includes/class-wporg-gp-translation-events-translation-listener.php

```diff
@@ -50,68 +56,123 @@ function ( $translation, $translation_before ) {
 		);
 	}
 
-	private function handle_action( GP_Translation $translation, int $user_id, string $action, DateTime $happened_at ): void {
-		// Get events that are active when the action happened, for which the user is registered for.
-		$active_events = $this->get_active_events( $happened_at );
-		$events        = $this->select_events_user_is_registered_for( $active_events, $user_id );
-
-		/** @var GP_Translation_Set $translation_set */
-		$translation_set = ( new GP_Translation_Set )->find_one( [ 'id' => $translation->translation_set_id ] );
-		global $wpdb;
-
-		foreach ( $events as $event ) {
-			// A given user can only do one action on a specific translation.
-			// So we replace instead of insert, which will keep only the last action.
-			$wpdb->replace(
-				self::ACTIONS_TABLE_NAME,
-				[
-					// start primary key
-					'event_id'       => $event->ID,
-					'user_id'        => $user_id,
-					'translation_id' => $translation->id,
-					// end primary key
-					'action'         => $action,
-					'locale'         => $translation_set->locale,
-					'happened_at'    => $happened_at->format( 'Y-m-d H:i:s' ),
-				]
-			);
+	private function handle_action( GP_Translation $translation, int $user_id, string $action, DateTimeImmutable $happened_at ): void {
+		try {
+			// Get events that are active when the action happened, for which the user is registered for.
+			$active_event_ids = $this->get_active_events( $happened_at );
+			$event_ids        = $this->select_events_user_is_registered_for( $active_event_ids, $user_id );
+
+			/** @var GP_Translation_Set $translation_set */
+			$translation_set = ( new GP_Translation_Set )->find_one( [ 'id' => $translation->translation_set_id ] );
+			global $wpdb;
+
+			foreach ( $event_ids as $event_id ) {
+				// A given user can only do one action on a specific translation.
+				// So we replace instead of insert, which will keep only the last action.
+				$wpdb->replace(
+					self::ACTIONS_TABLE_NAME,
+					[
+						// start primary key
+						'event_id'       => $event_id,
+						'user_id'        => $user_id,
+						'translation_id' => $translation->id,
+						// end primary key
+						'action'         => $action,
+						'locale'         => $translation_set->locale,
+						'happened_at'    => $happened_at->format( 'Y-m-d H:i:s' ),
+					]
+				);
+			}
+		} catch (Exception $exception) {
+			error_log($exception);
 		}
 	}
 
 	/**
-	 * @return WP_Post[]
+	 * @return int[]
+	 * @throws Exception
 	 */
-	private function get_active_events( DateTime $at ): array {
-		return get_posts(
+	private function get_active_events( DateTimeImmutable $at ): array {
+		$cache_entries = $this->active_events_cache->get();
```

Comment:
Can we just rename this to 
```suggestion
		$entries = $this->active_events_cache->get();
```